### PR TITLE
PGN

### PR DESCRIPTION
**Pull request by suprajasridhara**
Pull request for boilerplate code of PGN with the following changes:

- add PGN as a new service type and read mapping service entries from topology files
- define the configuration to start the service
- initialise the service with a messenger and database
- start service

### CODE REVIEW

File: go/pkg/pgn/config/configtest/configtest.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package configtest
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/configtest/configtest.go:15]: Please also add the copyright notice to this file

File: go/pkg/pgn/config/config.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	// Port to listen on (required)
+	Port uint16 `toml:"port,omitempty"`
+
+	// IA the local IA (required)
+	IA addr.IA `toml:"isd_as,omitempty"`
+
+	//CfgDir directory to read crypto keys from (required)
+	CfgDir string `toml:"cfg_dir,omitempty"`
+
+	//Db to store pgn cfg data (default ./pgn.db will be created or read from)
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/config.go]:
```suggestion
	//Db to store PNG cfg data (default ./pgn.db will be created or read from)
```

File: go/pkg/pgn/config/config.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	DefaultDb = "./pgn.db"
+)
+
+type Config struct {
+	General  env.General `toml:"general,omitempty"`
+	Features env.Features
+	Logging  log.Config       `toml:"log,omitempty"`
+	Metrics  env.Metrics      `toml:"metrics,omitempty"`
+	Sciond   env.SCIONDClient `toml:"sciond_connection,omitempty"`
+	Pgn      PgnConf          `toml:"pgn,omitempty"`
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/config.go]:
```suggestion
	PGN      PGNConf          `toml:"pgn,omitempty"`
```

File: go/pkg/pgn/config/config.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	// IA the local IA (required)
+	IA addr.IA `toml:"isd_as,omitempty"`
+
+	//CfgDir directory to read crypto keys from (required)
+	CfgDir string `toml:"cfg_dir,omitempty"`
+
+	//Db to store pgn cfg data (default ./pgn.db will be created or read from)
+	Db string `toml:"db,omitempty"`
+
+	//QUIC address to listen to quic IP:Port (required)
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/config.go]:
```suggestion
	//QUIC address to listen to QUIC IP:Port (required)
```

File: go/pkg/pgn/config/config.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		return serrors.New("ip must be set")
+	}
+	if cfg.Port == 0 {
+		return serrors.New("Port must be set")
+	}
+	if cfg.CfgDir == "" {
+		return serrors.New("pgn cfg_dir should be set")
+	}
+	if cfg.QUICAddr == "" {
+		return serrors.New("quic addr should be set")
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/config.go]:
```suggestion
		return serrors.New("QUIC addr should be set")
```

Use consistently either the struct field name or the toml config key.
Also applies to the PLN PR.

File: go/pkg/pgn/config/sample.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+const idSample = "pgn1"
+
+const pgnSample = `
+# ID of the PGN. (required)
+id = "%s"
+# IP to listen on (required)
+ip = "127.0.0.65"
+# Port to listen on (required)
+port = 3009
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/sample.go]: >`port = 3009`

Use different default ports in the sample configs for MS, PLN, PGN. One might use the sample configs for integration tests and local testing, which then port clash between the services.

File: go/pkg/pgn/config/config.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		&cfg.Pgn,
+	)
+}
+
+var _ config.Config = (*PgnConf)(nil)
+
+type PgnConf struct {
+	// ID of the PGN (required)
+	ID string `toml:"id,omitempty"`
+
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/config.go]: Remove the empty newlines here and below

File: go/pkg/pgn/config/config.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+const (
+	DefaultDb = "./pgn.db"
+)
+
+type Config struct {
+	General  env.General `toml:"general,omitempty"`
+	Features env.Features
+	Logging  log.Config       `toml:"log,omitempty"`
+	Metrics  env.Metrics      `toml:"metrics,omitempty"`
+	Sciond   env.SCIONDClient `toml:"sciond_connection,omitempty"`
```
Comment (FR4NK-W) [inline go/pkg/pgn/config/config.go]: >`Sciond   env.SCIONDClient `toml:"sciond_connection,omitempty"``

Try to follow the recommendations for the naming conventions from https://scion.docs.anapaya.net/en/latest/contribute/go-style.html

File: go/pgn/internal/pgncmn/common.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+func Init(cfg pgnconfig.PgnConf, sdCfg env.SCIONDClient, features env.Features) error {
+	router, err := infraenv.NewRouter(cfg.IA, sdCfg)
+	if err != nil {
+		return serrors.WrapStr("Error in Init mscmn", err)
+	}
+	nc := infraenv.NetworkConfig{
+		IA:                    cfg.IA,
+		Public:                &net.UDPAddr{IP: cfg.IP, Port: int(cfg.Port)},
+		SVC:                   addr.SvcPGN,
```
Comment (FR4NK-W) [inline go/pgn/internal/pgncmn/common.go:43]: Please lint the files in these PRs, at least gofmt and goimports

**Review by FR4NK-W — CHANGES_REQUESTED** [review]

File: go/pkg/pgn/config/configtest/configtest.go (hunk 1 above)
Comment (suprajasridhara) [reply go/pkg/pgn/config/configtest/configtest.go:15]: done

File: go/pgn/internal/pgncmn/common.go (hunk 9 above)
Comment (suprajasridhara) [reply go/pgn/internal/pgncmn/common.go:43]: This I had done, is there something specific that I am missing? I don't see any problems reported

Comment (FR4NK-W) [reply go/pgn/internal/pgncmn/common.go:43]: My bad, I had an old gofmt on the path from the debian package install of Go for Ubuntu 18.04:
```
circleci@golang:~/project/scion$ diff <(cat go/pgn/internal/pgncmn/common.go) <(/usr/lib/go-1.10/bin/gofmt go/pgn/internal/pgncmn/common.go)
41,43c41,43
<               IA:                    cfg.IA,
<               Public:                &net.UDPAddr{IP: cfg.IP, Port: int(cfg.Port)},
<               SVC:                   addr.SvcPGN,
---
>               IA:     cfg.IA,
>               Public: &net.UDPAddr{IP: cfg.IP, Port: int(cfg.Port)},
>               SVC:    addr.SvcPGN,
circleci@golang:~/project/scion$ diff <(cat go/pgn/internal/pgncmn/common.go) <(gofmt go/pgn/internal/pgncmn/common.go)
circleci@golang:~/project/scion$
```

I should have checked the actual output, there wasn't enough context to see that the current formatting is indeed correct.

**Pull request merged**